### ISSUE TITLE
on event push without PRs fallback to 0

### ISSUE THREAD

**Issue by fulgas** [DevSecTop/TF-via-PR#354]
#### Is your feature request related to a problem

When there is a push to the main branch and there is no PR associated with it, the PR Number is null. 

#### Describe the solution you'd like

Is it possible to use a fallback to 0 ?

**Comment by rdhar** [comment]
Hey @fulgas, TF-via-PR explicitly supports both `pull_request` and `push` event triggers, shown in the [Readme](https://github.com/DevSecTop/TF-via-PR?tab=readme-ov-file#usage) and [this example workflow](https://github.com/DevSecTop/TF-via-PR/blob/main/.github/examples/pr_push_auth.yaml).

This includes fetching the PR number associated with a commit, even from `push` trigger.

What's more, TF-via-PR also has a [fallback of `0`](https://github.com/DevSecTop/TF-via-PR/blob/main/.github/examples/pr_push_auth.yaml) if it can't determine the PR number from [various query methods](https://github.com/DevSecTop/TF-via-PR/blob/main/action.yml#L77-L87).

Are you able to share a little more about your workflow, with (obfuscated) snippets to aid troubleshooting.

**Comment by silnels-greenlane** [comment]
I'm running a `plan` when i `push` directly to the main branch.

This is the identifier for the file `terraform.null...chdir...tf.plan`

**Comment by fulgas** [comment]
I'm running a `plan` when i `push` directly to the main branch.

This is the identifier for the file `terraform.null...chdir...tf.plan`. 

The `${pr_number}` is `null` and since it fall into the` if [[ "$GITHUB_EVENT_NAME" == "push" ]];` im assuming that it didnt found any PR associated when the commit.

**Comment by rdhar** [comment]
Hmm, that doesn't sit right. Could you:

1. Bump up to the latest version `devsectop/tf-via-pr@v12.0.5` and share the new artifact name.
1. Confirm if there is any commit present on the PR branch?

**Comment by silnels-greenlane** [comment]
This is not your typical workflow as you showed on the examples.

I have 2 workflows, one for CI which runs on PR and produces the tf plan, and another workflow for CD that actually applies the plan.

The difference here is that besides that i am allowing to push directly to the main branch even when **no** PR is associated.

**Comment by fulgas** [comment]
This is not your typical workflow as you showed on the examples.

I have 2 workflows, one for CI which runs on PR and produces the tf plan, and another workflow for CD that actually applies the plan.

The difference here is that besides that i am allowing to push directly to the main branch even when **no** PR is associated.

**Comment by fulgas** [comment]
The result is the same: `terraform-null-3eee8c84341b95e00d4d4dd1307a8268.tfplan`

I can try to play with `label-pr` and `comment-pr` to `none` when a know that the push is not related to a PR on anyway.

**Comment by fulgas** [comment]
> gh api /repos/${GITHUB_REPOSITORY}/commits/${GITHUB_SHA}/pulls

This call returns `[]` which is true and a validate response in this case.

**Comment by rdhar** [comment]
Admittedly, I haven't yet run into the use case of having a CI workflow _as well as_ another workflow which circumvents it altogether.

I'm afraid neither `label-pr` nor `comment-pr` will help in this case, since the PR number is a crucial component of TF-via-PR.

Before working on a solution, could you share more details around pipeline flow and expectations?

- Do you have a `pull_request` trigger for the initial plan file generation and storage?
- For `push` without, do you expect to retrieve and apply the plan file from a certain PR?
- Or do you intend on `apply -auto-approve` approach instead?

**Comment by fulgas** [comment]
This pipeline will run a `plan` then an `apply`.

> * Do you have a `pull_request` trigger for the initial plan file generation and storage?

No PR

> * Upon `push` without PR, do you expect to retrieve and apply the plan file from a certain PR?

No

> * Or do you intend on a more direct `apply -auto-approve` approach instead?

No auto approve.

**Comment by fulgas** [comment]
> I'm afraid neither `label-pr` nor `comment-pr` will help in this case, since the PR number is a crucial component of TF-via-PR.

Works as i expected if i apply on the `plan` and `apply`

```
comment-pr: "none"
label-pr: false
```

I can live with the `pr_number` being `null`.

**Comment by rdhar** [comment]
No way, you figured out a resolution to this yourself?! Is there anything else to help with?

**Comment by fulgas** [comment]
All good. Thk for the help :)

**Comment by rdhar** [comment]
Thank _you_ for demonstrating a new use case! If the current identifier-naming strategy doesn't work out, feel free to drop us another message—happy to amend based on feedback.

**Comment by rdhar** [comment]
Although, you've managed to solve this, others (i.e., me) might've been stuck.

And given the simplicity—just push to main—I think this use case should be accounted for _without_ resorting to `null` combined with `comment-pr: none` and `label-pr: false`.

As per your suggestion, the PR number now falls back on `0` in this scenario, and no further need to disable `comment-pr` or `label-pr`.

Happy to share this has been shipped with [v12.0.7](https://github.com/DevSecTop/TF-via-PR/releases/tag/v12.0.7), where your contribution has been credited!

**Comment by tenpaiyomi** [comment]
@rdhar adding another few use case here that I've run into for this particular particular case, but I have a scenario where I am triggering a github action from an `issue_comment` `created` event to create a new targeted plan.

The tl;dr is that we sometimes want to be able to manually plan and apply against a PR before it is merged into our `main` branch, so I have set up a workflow that reacts to `/tofu-plan` and `/tofu-apply` comments to manually run the workflow and do a `plan` or `apply`.  I've got it actually running the workflow but noticed that it wasn't creating the comment in the PR.  Looking through the code I noted that [the identifier step](https://github.com/OP5dev/TF-via-PR/blob/main/action.yml#L78-L89) works on the basis of assuming we are going to be working from a `push` or `merge_group` event, otherwise it falls back to trying to get the event number from `github.event.number` which, in the case of the `issue_comment`  `created` event doesn't exist.

It seems like the easiest potential way to solve this use case would be to just allow manually passing in a `pr-number` value and updating the identifier step to be something along the lines of

```
        if [[ -n "${{ inputs.pr-number }}" ]]; then
          pr_number=${{ inputs.pr-number }}
        elif [[ "$GITHUB_EVENT_NAME" == "push" ]]; then
          # List PRs associated with the commit, then get the PR number from the head ref or the latest PR.
          associated_prs=$(gh api /repos/{owner}/{repo}/commits/${GITHUB_SHA}/pulls --header "$GH_API" --method GET --field per_page=100)
          pr_number=$(echo "$associated_prs" | jq --raw-output '(.[] | select(.head.ref == env.GITHUB_REF_NAME) | .number) // .[0].number // 0')
        elif [[ "$GITHUB_EVENT_NAME" == "merge_group" ]]; then
          # Get the PR number by parsing the ref name.
          pr_number=$(echo "${GITHUB_REF_NAME}" | sed -n 's/.*pr-\([0-9]*\)-.*/\1/p')
        else
          # Get the PR number from the event payload or fallback on 0.
          pr_number=${{ github.event.number || 0 }}
        fi
        echo "pr=$pr_number" >> "$GITHUB_OUTPUT"
```

And in the `inputs` section, just adding a

```
  pr-number:
    default: ""
    description: "Specify a PR number for the executed workflow"
    required: false
```

**Comment by rdhar** [comment]
Happy to share this has been shipped with [v13.0.1](https://github.com/OP5dev/TF-via-PR/releases/tag/v13.0.1), where your contribution has been credited!

Please consider ⭐ this project, if you find it useful.

[![GitHub repository stargazers](https://img.shields.io/github/stars/devsectop/tf-via-pr)](https://github.com/devsectop/tf-via-pr "Become a stargazer.")

---

Hi @tenpaiyomi, first off, thank you for taking the time to dig into the code and suggesting solutions: the gold standard for feedback! As it happens, it's a [one-line change](https://github.com/OP5dev/TF-via-PR/blob/main/action.yml#L87) to account for this event trigger automatically.

Though, a small word of caution: `issue_comment` is a pain in the butt.

Earlier version of this Action used to be solely comment-triggered for an Atlantis-like user experience, which sounds great on paper but turned out to be a nightmare to debug. The biggest drawback? Turns out, `issue_comment` event _always_ bases its workflow from the main branch instead of the PR one. Meaning that changes to the workflow have to be tested by merging to main directly.

Worse still, `issue_comment` doesn't come with half the metadata payload that's available from `pull_request` or `push` events. And that's nothing to say about new comments adding noise to the PR conversation for each trigger. It's a real shame that GitHub hasn't shown it love, because `issue_comment` trigger has so much potential.

Going back to your use-case, I'm happy to confirm that some of our teams ran into the same requirement and found a novel approach to trigger the workflow manually: with PR labels!

More specifically, `pull_request` event has an underrated `labeled` activity type which allows you to trigger a workflow based on the PR label. Why is this better than `issue_comment`?

- Avoids noisy clutter of "trigger" comments in the PR conversation thread.
- No chance of typos since you select from a dropdown of existing labels.
  - The label can removed automatically by the workflow so it can be re-triggered repeatedly.
- Since it's a sub-type of `pull_request`, it has access to all the usual event payload data.

That's all well and good, but the poof is in the pudding. As such, we've just shared a [functional workflow example](https://github.com/OP5dev/TF-via-PR/blob/main/.github/examples/pr_self_hosted.yaml#L10) which runs `plan` and `apply` as usual, in addition to triggering a `plan` whenever "tf-plan" PR label is attached. Keen to know if this is a good fit for you!

---

As an aside, I just clocked [your fork here](https://github.com/OP5dev/TF-via-PR/compare/main...StoatLabs:TF-via-PR:main#diff-1243c5424efaaa19bd8e813c5e6f6da46316e63761421b3e5f5c8ced9a36e6b6R360), and had a couple Qs to better understand it.

1. What's the context of `identifier-prefix`? As in, what does it add to the identifier? 
1. If you wanna amend the PR comment output, you could hook into the "`<!-- placeholder-4 -->`" string. [Here's an example](https://github.com/3ware/aws-network-speciality/blob/main/.github/workflows/terraform-ci.yaml#L137-L138) from out in the wild.

**Comment by tenpaiyomi** [comment]
First off, @rdhar, thank you so much for the quick response and turnaround.  I'm used to offering a thought or suggestion and then not hearing anything for weeks or months, if at all, so it was incredibly surprising and refreshing.

> Hi [@tenpaiyomi](https://github.com/tenpaiyomi), first off, thank you for taking the time to dig into the code and suggesting solutions: the gold standard for feedback! As it happens, it's a [one-line change](https://github.com/OP5dev/TF-via-PR/blob/main/action.yml#L87) to account for this event trigger automatically.

I noticed this after you made the change and had a "huh, yeah, that's much smarter" moment, which is why I reverted my change in the fork I made and went with yours.

> Though, a small word of caution: `issue_comment` is a pain in the butt.
> 
> Earlier version of this Action used to be solely comment-triggered for an Atlantis-like user experience, which sounds great on paper but turned out to be a nightmare to debug. The biggest drawback? Turns out, `issue_comment` event _always_ bases its workflow from the main branch instead of the PR one. Meaning that changes to the workflow have to be tested by merging to main directly.
> 
> Worse still, `issue_comment` doesn't come with half the metadata payload that's available from `pull_request` or `push` events. And that's nothing to say about new comments adding noise to the PR conversation for each trigger. It's a real shame that GitHub hasn't shown it love, because `issue_comment` trigger has so much potential.

I had actually run into this caveat early on and had to think about how to work with it.  The way I wound up doing so was fairly straightforward.  I have code that works based on various `pull_request` types in one workflow and then another workflow for the `issue_comment` `created` type.  Each of those has a `uses` block that calls a shared bit of code for the workflow and pass some inputs, specifically in this case a `ref` input.  The `pull_request` types pass `${{ github.context.ref }}`, while the `issue_comment` type passes `refs/pull/${{ github.event.issue.number }}/head`.  This helped to ensure I was always working on the appropriate branch code.

> Going back to your use-case, I'm happy to confirm that some of our teams ran into the same requirement and found a novel approach to trigger the workflow manually: with PR labels!
> 
> More specifically, `pull_request` event has an underrated `labeled` activity type which allows you to trigger a workflow based on the PR label. Why is this better than `issue_comment`?
> 
>     * Avoids noisy clutter of "trigger" comments in the PR conversation thread.
> 
>     * No chance of typos since you select from a dropdown of existing labels.
>       
>       * The label can removed automatically by the workflow so it can be re-triggered repeatedly.
> 
>     * Since it's a sub-type of `pull_request`, it has access to all the usual event payload data.
> 
> 
> That's all well and good, but the poof is in the pudding. As such, we've just shared a [functional workflow example](https://github.com/OP5dev/TF-via-PR/blob/main/.github/examples/pr_self_hosted.yaml#L10) which runs `plan` and `apply` as usual, in addition to triggering a `plan` whenever "tf-plan" PR label is attached. Keen to know if this is a good fit for you!

I actually _really_ like this idea and, as you noted and I found out with my testing of the comment flow, would definitely help with keeping the noise clutter down.  As is now, the way we are utilizing the flow a little bit already makes use of labels.  Specifically, we wanted a situation where we could have some PRs target specific instances of our infrastructure so we could deploy out to those instances for testing as part of our pipeline, and in such I created prefixed labels to correspond to the specific instances (i.e. `target:foo`, `target:foo-staging`, `target:bar-staging`, etc.) and then have those labels read in as part of the workflow.  Any that are read are then modified to drop `target:` and prepend `module.`, then joined together into a single string such that I get something like `module.foo-staging,module.bar-staging` which then gets passed to the `arg-target` argument.

The big sort of caveat / question mark here though is what sort of changes would happen to the infrastructure as a whole when it isn't targeting specific instances, as well as how to handle the situation of merging all the code into `main` and wanting to apply it to all the infrastructure.  That leads to the next part...

> As an aside, I just clocked [your fork here](https://github.com/OP5dev/TF-via-PR/compare/main...StoatLabs:TF-via-PR:main#diff-1243c5424efaaa19bd8e813c5e6f6da46316e63761421b3e5f5c8ced9a36e6b6R360), and had a couple Qs to better understand it.
> 
>     1. What's the context of `identifier-prefix`? As in, what does it add to the identifier?

The main idea here was that I wanted PRs to be tracking both a full plan as well as any targeted plans, notably because when a pull request is merged into `main`, as noted above, the expectation is that it would be applied to all the instances, which in turn would mean that it would be looking for a tfplan.  If the plan it had been working with before the merge was a targeted plan, then the merge would result in the apply also being for the targeted changes, so I added the `identifier-prefix` idea such that I could run the `plan` command twice on my PRs.  One run uses the prefix of `full`, and generates a `tofu-PR#-full-xxxxxxxxx.tfplan`, whereas the other has a prefix of `targeted` and, if any targeting labels are applied to the PR, generates a `tofu-PR#-targeted-xxxxxxxxx.tfplan`.

This, coupled with a few conditionals pass through as inputs on the workflow, gives me the flexibility to know that my `/tofu-apply` comment will make the appropriate changes only on the targeted instances because it will download and apply the `targeted` prefixed plan.

There are a few little bits that I still need to work through and improve, such as the point that right now there is no way to actually apply a PR plan if it does not have any targeted instances.  That was done more so as an initial safety measure to ensure a PR isn't accidentally given an apply command that affects everything, but there are some valid theoretical scenarios where such would be desired.  There is also the scenario where, when a PR does get merged in, it _will_ do the full infrastructure apply, which means if there are some PRs that are assigned to some other instances, they will have to have their apply run again to redeploy their branch out.  Fairly minor caveats that I'm just having to figure out how best to handle with the tools that Github Actions affords us.

>     2. If you wanna amend the PR comment output, you could hook into the "`<!-- placeholder-4 -->`" string. [Here's an example](https://github.com/3ware/aws-network-speciality/blob/main/.github/workflows/terraform-ci.yaml#L137-L138) from out in the wild.

Ahaha, I am guessing you noted the several commits I made trying to get the formatting on that set up.  I was having issues with newlines before I realized that, within the `summary` block, the `<br/>` has to have extraneous space around it to actually work properly.

This is the overall output for a test branch

![Image](https://github.com/user-attachments/assets/5760205d-b390-4fa4-9921-6706f1c4aa88)
![Image](https://github.com/user-attachments/assets/e38ba8b0-0714-49db-a9da-ebc7e027b0ce)

**Comment by rdhar** [comment]
Away from my laptop just now, so short on formatting:

1. You're not separately running "init" and then "plan" commands in different steps, are you? If so, you can drop the "init" step altogether. The "plan" already includes "init". The only reason "init" is supported by itself is for some specific CI use-cases where folks want to format/validate their config instead of running "plan" or "apply". The same goes for "apply", whereby "init" is included. Perhaps I should document this better. 
1. "_If the plan it had been working with before the merge was a targeted plan, then the merge would result in the apply also being for the targeted changes._" Mostly there. The Action doesn't blindly apply the last plan associated with a PR branch. It applies the last plan with the same matching args input from the branch. In other words, if the workflow is run with the arg inputs for a full plan, then the arg inputs for a targeted plan, then ends up being merged with the full plan's inputs, will lead to the full plan being applied while the targeted plan remains unused. The plan file naming convention is deterministic so the same inputs always results in the same output, which is how the right plan gets applied depending on arg inputs.
1. Super neat to hear `arg-target` being used! It's one of the args which gives me the heebie-jeebies and, admittedly, reading about the label-based module targeting did my heart rate no good! That being said, I get it. Some of our teams internally swear by labels for each and every aspect of their pipeline. Not been long since I saw labels to trigger plan/apply on specific TF workspaces for different dev/test/prod envs. Wild what GitHub Actions let us get away with these days.

**Comment by tenpaiyomi** [comment]
> Away from my laptop just now, so short on formatting:

Same, so please excuse any typos or lack of expansion. 

> 1. You're not separately running "init" and then "plan" commands in different steps, are you? If so, you can drop the "init" step altogether. The "plan" already includes "init". The only reason "init" is supported by itself is for some specific CI use-cases where folks want to format/validate their config instead of running "plan" or "apply". The same goes for "apply", whereby "init" is included. Perhaps I should document this better.

I am, but had noted it was redoing the init in the plan and apply step. I was planning on pulling this out, but it is why I had the validation step, since it was for the purpose of the format/validate.

> 1. "_If the plan it had been working with before the merge was a targeted plan, then the merge would result in the apply also being for the targeted changes._" Mostly there. The Action doesn't blindly apply the last plan associated with a PR branch. It applies the last plan with the same matching args input from the branch. In other words, if the workflow is run with the arg inputs for a full plan, then the arg inputs for a targeted plan, then ends up being merged with the full plan's inputs, will lead to the full plan being applied while the targeted plan remains unused. The plan file naming convention is deterministic so the same inputs always results in the same output, which is how the right plan gets applied depending on arg inputs.

I'm not 100% sure that is fully accurate. I've no doubt it might be the intention, but if you look at the screenshot I shared, outside of the prefix I added the filenames are identical. This was what initially lead me to adding in the prefix. I _intentionally_ wanted to create both a full plan, always, and then also a targeted plan if applicable. That way if I did the apply from a merge, the full plan would be available in the artifacts, but if I did an apply from the comments and it was targeted, that would also be available in the artifacts. When I am doing the plan, I am specifically doing one without the target arg with the full identifier prefix and then, if there are any targets, I do a separate plan with the target args. However, I found that the full tfplan was being created and then overwritten when I did another tfplan with the target arguments, even though their output was different.

> 1. Super neat to hear `arg-target` being used! It's one of the args which gives me the heebie-jeebies and, admittedly, reading about the label-based module targeting did my heart rate no good! That being said, I get it. Some of our teams internally swear by labels for each and every aspect of their pipeline. Not been long since I saw labels to trigger plan/apply on specific TF workspaces for different dev/test/prod envs. Wild what GitHub Actions let us get away with these days. 

Yeah, it's one of the things I wish was a little bit better handled in terraform/tofu. There are a lot of times I've had to deploy out to many instances that are all production, but then there are also some that are staging or qe, but they all wind up existing in the same account and often have some generally common infra amongst them. Because it doesn't make sense to do a plan and apply for 20+ instances to do the same thing, I wound up making them module instances with any applicable vars to differentiate them. This also allows me to target for applications. FWIW, I have been fairly defensive in how I've handled the labels for the targeting. I have a separate workflow that ensures the label usage is unique amongst PRs, while also adding a label that explicitly prohibits any infrastructure applying until another plan is generated and confirmed.

**Comment by tenpaiyomi** [comment]
I went back and looked at the code again for the identifier and noted that it only uses a few particular args for uniqueness. The target, in this case, isn't one, which explains why my identifiers were not unique.

**Comment by rdhar** [comment]
##### Consolidate init+plan/apply steps.

Ah gotcha. If you're doing fmt/validate in a separate pre-job to plan, then that's grand. Otherwise, you can totally include `format: true` and `validate: true` inputs to the plan step, and it'll carry out those checks as part of preparing the plan, so you don't have to spend time re-doing init again.

##### Same artifact name despite different inputs.

Shoot, you got me. Taking a closer look at the screenshots, I see what you mean about the identical naming (and also missing timestamp from the PR comment, but that's not as important (no thanks to `issue_comment`'s sparse payload)).

##### Not so unique after all.

As it stands, the "unique" identifier takes into account a handful of curated inputs so as to treat TF projects separately (e.g., different `arg-var-file`) but also the same (e.g., different `arg-parallelism`, `arg-lock`, or `arg-auto-approve`). This stands at:

- `chdir` (i.e., working directory)
- `backend-config`
- `var-file`
- `var`
- `workspace`
- `destroy`
- `tool` (i.e., tofu or terraform)
- PR number.

Thanks to your feedback, I reckon we're due a couple new additions to the mix:

- `target`
- `replace`

Should be able to a cut a new patch-release in a couple hours to ship this change.

**Comment by rdhar** [comment]
Thanks once again for your feedback (and ridiculously generous sponsorship), @tenpaiyomi!

Happy to share both fixes for the `arg-target` identifier and missing `issue_comment` timestamp have shipped with [v13.0.2](https://github.com/OP5dev/TF-via-PR/releases/tag/v13.0.2), where your contribution has been credited!

On a related tangent, previous contributor @p4perf4ce had [mentioned](https://github.com/OP5dev/TF-via-PR/pull/272#issuecomment-2280188272) leveraging [github/branch-deploy](https://github.com/github/branch-deploy) for IssueOps via GitHub Actions. I suspect you've grown past this already, given your interlinked/reusable workflows, but maybe worth a glance for comments-based triggers. With that said, I still prefer PR labels for this purpose!

**Comment by rdhar** [comment]
Been a few days past, and keen to follow up on if the separate full/target plan files have worked out for you, and your progress with issue_comment or PR label triggers, @tenpaiyomi?
